gtd_generative_proof_machine/inc/generative_restriction.h

```diff
@@ -0,0 +1,60 @@
+#ifndef GENERATIVE_RESTRICTION_H
+#define GENERATIVE_RESTRICTION_H
+
+#include "common.h"
+#include "physical_graph.h"
+#include "generative_proof_machine.h"
+
+typedef struct GenerativeRestriction GenerativeRestriction;
+typedef struct intNode intNode;
+typedef struct PathNode PathNode;
+
+typedef struct PathNode
+{
+    int *path;
+    int *pathBool;
+    int numElems;
+    PathNode *next;
+} PathNode;
+
+typedef struct RestrictionResult
+{
+    int modified;
+    int contradictionFound;
+} RestrictionResult;
+
+typedef struct RestrictionParameters
+{
+    int numIntParams;
+    int *intParams;
+    GenerativeProofMachine *machine;
+} RestrictionParameters;
+
+RestrictionParameters *initialize_restriction_parameters(void);
+void destroy_restriction_parameters(RestrictionParameters *params);
+PathNode *find_k_paths(Graph *graph, int k);
+GenerativeRestriction *create_restriction_object(RestrictionResult *(*condition)(Graph *graph, RestrictionParameters *params), RestrictionParameters *params);
+int delete_restriction_object(GenerativeRestriction *restriction);
+RestrictionResult *validate_restriction(Graph *graph, GenerativeRestriction *restriction);
+
+GenerativeRestriction *create_max_degree_restriction(int maxVertices, GenerativeProofMachine *machine);
```

Comment:
consider introducing restriction type (best option is to move to common and reuse FactType enum, not every type needs to be supported both in generative machine and fact tree), then use one public constructor for GenerativeRestriction class with type argument

Reply:
it is not a simple fix. I will put it on todo list for the next pr